exporter/elasticsearchexporter/internal/lru/lruset.go

```diff
@@ -0,0 +1,74 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package lru // import "github.com/open-telemetry/opentelemetry-collector-contrib/exporter/elasticsearchexporter/internal/lru"
+
+import (
+	"time"
+
+	"github.com/cespare/xxhash"
+	"github.com/elastic/go-freelru"
+	"go.opentelemetry.io/ebpf-profiler/libpf/xsync"
+)
+
+type void struct{}
+
+func stringHashFn(s string) uint32 {
+	return uint32(xxhash.Sum64String(s))
+}
+
+// LockedLRUSet is the interface provided to the LRUSet once a lock has been
+// acquired.
+type LockedLRUSet interface {
+	// CheckAndAdd checks whether the entry is already stored in the cache, and
+	// adds it.
+	// It returns whether the entry should be excluded, as it was already present
+	// in cache.
+	CheckAndAdd(entry string) bool
+}
+
+// LRUSet is an LRU cache implementation that allows acquiring a lock, and
+// checking whether specific keys have already been stored.
+type LRUSet struct {
+	syncMu *xsync.RWMutex[*freelru.LRU[string, void]]
+}
+
+func (l *LRUSet) WithLock(fn func(LockedLRUSet) error) error {
+	if l == nil || l.syncMu == nil {
+		return fn(nilLockedLRUSet{})
+	}
```

Comment:
Is this a valid case? Should the function be called here without any mutex mechanism occurring, or should an error be returned instead? (I'm wondering if perhaps this hides an error from the user when they should be notified they didn't initialize the LRUSet properly.)

Reply:
It's not really an issue indeed, since the case can only happen in tests, when we can manually create an LRUSet.
I do agree that returning an error simplifies the logic though. I've done that.